gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+
+    def __init__(self, sc):
+        super(ReportStateTask, self).__init__(cfg.CONF)
+        self._sc = sc
+        pulse = loopingcall.FixedIntervalLoopingCall(
+            self.run_periodic_tasks, None, None)
+        pulse.start(interval=cfg.CONF.reportstate_interval, initial_delay=None)
+
+    @oslo_periodic_task.periodic_task(spacing=5)
+    def report_state(self, context):
+        LOG.debug(_("Report state task invoked !"))
+        self._sc.report_state()
+
+
+class PeriodicTask(oslo_periodic_task.PeriodicTasks):
+
+    def __init__(self, sc):
+        super(PeriodicTask, self).__init__(cfg.CONF)
+        self._sc = sc
+        pulse = loopingcall.FixedIntervalLoopingCall(
+            self.run_periodic_tasks, None, None)
+        pulse.start(interval=cfg.CONF.periodic_interval, initial_delay=None)
+
+    @oslo_periodic_task.periodic_task(spacing=1)
+    def periodic_sync_task(self, context):
+        LOG.debug(_("Periodic sync task invoked !"))
+        self._sc.timeout()
+
+
+class Event(object):
+
+    def __init__(self, **kwargs):
+        self.serialize = kwargs.get(
+            'serialize') if 'serialize' in kwargs else False
+        self.binding_key = kwargs.get(
+            'binding_key') if 'binding_key' in kwargs else None
+        self.id = kwargs.get('id')
+        self.key = kwargs.get('key')
+        self.data = kwargs.get('data') if 'data' in kwargs else None
+        self.handler = kwargs.get('handler') if 'handler' in kwargs else None
+        self.poll_event = None  # Not to be used by user
+        self.worker_attached = None  # Not to be used by user
+        self.last_run = None  # Not to be used by user
+        self.max_times = -1  # Not to be used by user
+
+    def identify(self):
+        return "(id=%s,key=%s)" % (self.id, self.key)
+
+
+class EventCache(object):
+
+    def __init__(self, sc):
+        self._sc = sc
+        self._cache = core_queue.Queue(sc)
+
+    def rem(self, ev):
+        self._cache.remove([ev])
+
+    def rem_multi(self, evs):
+        self._cache.remove(evs)
+
+    def add(self, ev):
+        self._cache.put(ev)
+
+    def copy(self):
+        return self._cache.copy()
+
+
+class Serializer(object):
+
+    def __init__(self, sc):
+        self._sc = sc
+        '''
+        {'pid':{'binding_key':{'in_use':True, 'queue':[]}}}
+        '''
+        self._serializer_map = {}
+
+    def serialize(self, ev):
+        queued = False
+        LOG.debug(_("Serialize event %s" % (ev.identify())))
+        self._sc.lock()
+        if ev.worker_attached not in self._serializer_map:
+            self._serializer_map[ev.worker_attached] = {}
+        mapp = self._serializer_map[ev.worker_attached]
+        if ev.binding_key in mapp.keys():
+            queued = True
+            LOG.debug(_(
+                "There is already an event in progress"
+                "Queueing event %s" % (ev.identify())))
+
+            mapp[ev.binding_key]['queue'].append(ev)
```

Comment:
Yes, it is stored in the temp cache because some other related event is already being processed.
Please check, "class EventHandler --> _get()" method.